### ISSUE TITLE
`allowAutomaticSingleRunInference` breaks monorepo `compilerOptions.types`

### ISSUE THREAD

**Issue by aaronadamsCA** [typescript-eslint/typescript-eslint#4435]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

With `allowAutomaticSingleRunInference: true` enabled in a medium-sized React monorepo, `compilerOptions.types` no longer works as expected in one of our packages. My best guess is that the package is being linted using the wrong `tsconfig`.

**Repro**

With this config:

<details><summary><code>/packages/components/tsconfig.json</code></summary>

```jsonc
{
  "extends": "../../tsconfig.json",
  "include": ["./src/**/*"],
  "compilerOptions": {
    "types": ["segment-analytics"]
  }
}
```

</details>

And this line of code:

```ts
window.analytics.track(process.env["FOO"] ?? "foo");
```

**Expected behaviour**

Without `allowAutomaticSingleRunInference`, `@typescript-eslint/no-unsafe-member-access` gives us exactly the error we'd expect:

```
Unsafe member access .env on an `any` value
```

That's because I didn't include `"node"` in `compilerOptions.types`, so `process.env` is untyped.

**Actual behaviour**

With `allowAutomaticSingleRunInference: true`, it's a different story:

```
Unsafe member access .track on an `any` value
```

Oops. Somehow `process.env` is typed, but `window.analytics` is untyped. Both are incorrect; so clearly our `compilerOptions.types` isn't being respected.

No project references. We do use `paths` but they're not involved in this specific situation AFAIK. Our `tsconfig.json` inheritance is straightforward; I've included more of our config info below.

This reliably reproduces both locally and in CI, so I'm reasonably confident it isn't an environment issue.

Happy to hear any ideas RE: how we might create a standalone repro or help debug the issue. Access to our private repo could be an option, too.

**Additional Info**

<details><summary><code>/tsconfig.json</code></summary>

```jsonc
{
  "compilerOptions": {
    "baseUrl": "./",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "incremental": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "module": "ESNext",
    "moduleResolution": "Node",
    "noEmit": true,
    "noImplicitOverride": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "paths": {
      "@reach/router": ["./node_modules/@gatsbyjs/reach-router/es"],
      "@shopnishe/components": ["./packages/components/src"]
      // Our other internal packages
    },
    "skipLibCheck": true,
    "strict": true,
    "target": "ESNext"
  }
}
```

</details>

<details><summary><code>/.eslintrc.cjs</code></summary>

```js
module.exports = {
  extends: "./.eslintrc.base.cjs",
  overrides: [
    {
      files: ["*.ts", "*.tsx"],
      parserOptions: {
        allowAutomaticSingleRunInference: true,
        tsconfigRootDir: __dirname,
        project: ["./apps/*/tsconfig.json", "./packages/*/tsconfig.json"],
      },
      extends: [
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      rules: {
        // Our typed TypeScript rules
      },
    },
  ],
};
```

</details>

<details><summary><code>/.eslintrc.base.cjs</code></summary>

```js
module.exports = {
  root: true,
  extends: [
    "eslint:recommended",
    "plugin:react/recommended",
    "prettier", // Disables conflicting rules in above configs
    "plugin:react-hooks/recommended",
    "plugin:eslint-comments/recommended",
    "plugin:import/recommended",
    "plugin:jsx-a11y/strict",
  ],
  parser: "@babel/eslint-parser",
  parserOptions: {
    sourceType: "module",
  },
  env: {
    browser: true,
    es2020: true,
    node: true,
  },
  plugins: ["simple-import-sort"],
  rules: {
    // Our base rules
  },
  overrides: [
    {
      // Make sure ESLint can see every possible Node/TypeScript file
      files: ["*.cjs", "*.js", "*.jsx", "*.mjs", "*.ts", "*.tsx"],
    },
    {
      files: ["*.ts", "*.tsx"],
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:import/typescript",
      ],
      plugins: ["eslint-plugin-tsdoc"],
      rules: {
        // Our non-typed TypeScript rules
      },
    },
  ],
  settings: {
    // Our eslint-plugin-import settings
    react: {
      version: "detect",
    },
  },
};
```

</details>

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `5.9.0` |
| `TypeScript`                           | `4.5.4` |
| `node`                                 | `16.13.1` |

```
Environment Info:

Node version: v16.13.1
npm version: v8.1.2
Local ESLint version: v8.6.0 (Currently used)
Global ESLint version: Not found
Operating System: linux 4.19.128-microsoft-standard
```

**Comment by aaronadamsCA** [comment]
I managed to work around this in our codebase by moving to a single `tsconfig.json`, so it's not critical for us.

I tried to create an isolated repro, so far without success; I can't find the magic ingredient that (I think) causes this file to be assigned to the wrong program. I might give it another shot tomorrow.

**Comment by bradzacher** [comment]
Without a repro - we won't be able to investigate this, unfortunately. The TS compiler is big, complicated (and not our domain of expertise) - so without a concrete starting point I wouldn't even know where to begin!

The best bet would be to recreate your config in a new repo one file at a time until it breaks. The trouble with reproducing monorepo bugs is that often times it is some unique interdependency somewhere that causes them, which can be difficult to track down.
It's often not what you think!

You can use our debug logs to help investigate which file gets attached to what tsconfigs `DEBUG=typescript-eslint:*`

**Comment by aaronadamsCA** [comment]
@bradzacher Here you go!

https://github.com/aaronadamsCA/typescript-eslint-issue-4435

You can see the failed check here:

https://github.com/aaronadamsCA/typescript-eslint-issue-4435/runs/4796277237?check_suite_focus=true

Here's a full debug log from my local workspace:

```
$ DEBUG=typescript-eslint:* yarn eslint .
yarn run v1.22.15
$ /workspaces/nishe-web-apps/node_modules/.bin/eslint .
  typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(2) {
  '/workspaces/nishe-web-apps/apps/some-app/tsconfig.json',
  '/workspaces/nishe-web-apps/packages/components/tsconfig.json'
} +0ms
  typescript-eslint:typescript-estree:useProvidedProgram Retrieving ast for /workspaces/nishe-web-apps/apps/some-app/src/footer.tsx from provided program instance(s) +0ms
  typescript-eslint:typescript-estree:parser Detected single-run/CLI usage, creating Program once ahead of time for project: /workspaces/nishe-web-apps/apps/some-app/tsconfig.json +3ms
  typescript-eslint:parser:parser Resolved libs from program: [ 'lib' ] +0ms
  typescript-eslint:typescript-estree:parser parserOptions.project (excluding ignored) matched projects: Set(2) {
  '/workspaces/nishe-web-apps/apps/some-app/tsconfig.json',
  '/workspaces/nishe-web-apps/packages/components/tsconfig.json'
} +1s
  typescript-eslint:typescript-estree:useProvidedProgram Retrieving ast for /workspaces/nishe-web-apps/packages/components/src/index.tsx from provided program instance(s) +1s
  typescript-eslint:parser:parser Resolved libs from program: [ 'lib' ] +65ms
  typescript-eslint:eslint-plugin:utils:types Found an "error" any type +0ms
  typescript-eslint:eslint-plugin:utils:types Found an "error" any type +4ms
  typescript-eslint:eslint-plugin:utils:types Found an "error" any type +1ms

/workspaces/nishe-web-apps/packages/components/src/index.tsx
  10:5  error  Unsafe member access .track on an `any` value  @typescript-eslint/no-unsafe-member-access
  10:5  error  Unsafe call of an `any` typed value            @typescript-eslint/no-unsafe-call

✖ 2 problems (2 errors, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

As you can see, only one program is created; so it seems `packages/components/src/index.tsx` is linted using config `apps/some-app/tsconfig.json`.

**Comment by JoshuaKGoldberg** [comment]
@aaronadamsCA I think I found what the issue is: the `some-app/tsconfig.json` _is_ the "right" TSConfig to use, because it does include the `components/src/index.tsx` file!

In `typescript-estree`'s `getAstFromProgram` [[link](https://github.com/typescript-eslint/typescript-eslint/blob/d48429d97326545bb727f88ce9056270b1599a31/packages/typescript-estree/src/create-program/shared.ts#L114)], `currentProgram.getSourceFile(extra.filePath);` is used to get the source file (AST) for each linted file. That `currentProgram` is retrieved from an iterable created in `parseAndGenerateServices` [[link](https://github.com/typescript-eslint/typescript-eslint/blob/a02c63a264e5a6c3557468a8eff48d34ca2b718a/packages/typescript-estree/src/parser.ts#L499)].

If we get the list of all files in the program, we see that both the app file and the component file are included.

```plaintext
> currentProgram.getSourceFiles().map(sourceFile => sourceFile.fileName).filter(fileName => fileName.endsWith(".tsx"))
[
  0: '/home/josh/repos/typescript-eslint-issue-4435/packages/components/src/index.tsx'
  1: '/home/josh/repos/typescript-eslint-issue-4435/apps/some-app/src/footer.tsx'
]
```

Looking at the repo's sub-tsconfigs, they both `"include": ["./src/**/*"],` without setting a `baseUrl`. I _think_ that means they include everything under `src/` from their extended `baseUrl` of the root tsconfig's `./`. 

I tested this theory by added an intentionally wrong `export const wrong: number = "gotcha!"` in the **components** directory's `src/index.tsx` and then running `tsc` from the **apps/some-app** directory. Ignoring other errors, it did show up:

```ts
> ~/repos/typescript-eslint-issue-4435/apps/some-app$ tsc

../../packages/components/src/index.tsx:3:12 - error TS2322: Type 'string' is not assignable to type 'number'.

3 export var wrong: number = "hi";
             ~~~~~
```

I suspect what you'll want to do is configure your sub-tsconfigs to have their own `baseUrl`.

Closing this as an issue in user-land but please do post back if you think I'm wrong (which is not improbable; I don't have the strongest grasp on TSConfig extension nuances 😅).

**Comment by JoshuaKGoldberg** [comment]
OH, wait. I think I just recontextualized this internally - the program will have a source file for any file it includes, even if that file isn't included in the program. That includes `lib.d.ts` files, `@types/` and other `node_modules/` files, and cross-referenced files.

I wonder if we need a way to find the first program those include list includes a given file? Will investigate further. This is a fun one! 😄

**Comment by aaronadamsCA** [comment]
@JoshuaKGoldberg, that's very interesting! Great find.

I don't _think_ `baseUrl` is meant to affect `include`. Maybe that's the real bug here?

We set `baseUrl` only because it (unfortunately) must be specified when `paths` is set. I don't think it's meant to affect `include` behavior at all:

> ### [Include - `include`](https://www.typescriptlang.org/tsconfig#include)
>
> Specifies an array of filenames or patterns to include in the program. These filenames are resolved relative to the directory containing the `tsconfig.json` file.

I searched for any TypeScript issues/questions that mention both `baseUrl` and `include` and didn't find anything relevant, which might further support that the two settings should be independent.

I also pushed a commit to test whether `baseUrl` affects `include`, by adding a file at `/src/test.ts`. If I'm understanding your theory correctly, both programs would pick up that file, since it would match both of their `include` settings as calculated relative to their inherited `baseUrl` setting. But neither program seems to capture the file.

Hopefully you still find this issue "fun" after this added information 😅

**Comment by JoshuaKGoldberg** [comment]
_**OK!**_ ~5 months later, I think I actually have an update for you @aaronadamsCA! (I hope you're still working on the same original project!) ...This issue has been on my mind "fun" and also daunting since first posted in January. 🥲

I have more conviction now on the root of the issue indeed being that your `.eslintrc.cjs` says that _all TypeScript files_ may be have _all the TypeScript projects_ [[link](https://github.com/aaronadamsCA/typescript-eslint-issue-4435/blob/ae472237151693a7dadfab76cdbdbcde7639e34b/.eslintrc.cjs#L12):

```js
files: ["*.ts", "*.tsx"],
// ...
parserOptions: {
  allowAutomaticSingleRunInference: true,
  project: ["./apps/*/tsconfig.json", "./packages/*/tsconfig.json"],
  tsconfigRootDir: __dirname,
},
```

In short: we need a way to determine, given a `ts.Program` and file name `string`, whether the program includes that file name. We could then use that in `getAstFromProgram` to not bother getting a file's AST if it's not intentionally included by the program.

Two problems:

* I don't know the TypeScript API to ask for such a thing 😄. Asked in the TypeScript Discord: https://discord.com/channels/508357248330760243/640177429775777792/1006216471548010546
* This might present problems with code that expects being able to retrieve ASTs for referenced files outside of the programs?
    * If so: perhaps we need to first check for programs that intentionally include the file, then fall back to programs that don't...?

Experimenting now in https://github.com/typescript-eslint/typescript-eslint/pull/5445. Another maintainer should let me know soon if I'm going down a completely wrong path 😉 this is very much a learning task for me.

**Comment by JoshuaKGoldberg** [comment]
In the meantime, a workaround you can do is get more granular with specifying the `tsconfig.json` for source files. Instead of specifying an array of _all_ TSConfigs for all files, specify the proper TSConfig for each directory.

```js
module.exports = {
  root: true,
  overrides: [
    {
      files: ["*.ts", "*.tsx"],
      extends: [
        "plugin:@typescript-eslint/recommended",
        "plugin:@typescript-eslint/recommended-requiring-type-checking",
      ],
      // Instead of this...
      // parserOptions: {
      //   allowAutomaticSingleRunInference: true,
      //   project: ["./apps/*/tsconfig.json", "./packages/*/tsconfig.json"],
      //   tsconfigRootDir: __dirname,
      // },
    },
    // ...you can use this:
    {
      files: ["./apps/**/*.{ts,tsx}"],
      parserOptions: {
        allowAutomaticSingleRunInference: true,
        project: ["./apps/*/tsconfig.json"],
        tsconfigRootDir: __dirname,
      },
    },
    {
      files: ["./packages/**/*.{ts,tsx}"],
      parserOptions: {
        allowAutomaticSingleRunInference: true,
        project: ["./packages/*/tsconfig.json"],
        tsconfigRootDir: __dirname,
      },
    },
  ],
};
```

☝️ seemed to fix https://github.com/aaronadamsCA/typescript-eslint-issue-4435 locally for me. Can you verify whether this workaround fits for you @aaronadamsCA? 😄

**Comment by JoshuaKGoldberg** [comment]
It's been >1 month of no activity here. Since we have a workaround and it doesn't seem like TypeScript is built to support a more targeted solution, I'm going to go ahead and close this out. Thanks everyone!

Anybody, let me know if we should reopen and take action here

**Comment by aaronadamsCA** [comment]
Nah, that seems fair. Like I said from the start, we've switched to a single-`tsconfig.json` solution, and at this point I doubt we're looking back from that; and I'd agree with your statement that "it doesn't seem like TypeScript is built to support a more targeted solution".

I still think the repro is probably valid, but obviously there hasn't been a parade of people coming in with similar complaints since my original report at the start of the year. So I think we can safely move on :)

Thank you for investigating, though! If anything it was good that this "forced" us to a single tsconfig and all the benefits that brings.
